utilities/scale-json/src/lib.rs

```diff
@@ -0,0 +1,391 @@
+			}
+		)+
+	};
+}
+
+macro_rules! decode_to_hex {
+	(
+		$(
+			$name:ident, $t:ty $( | $convert:ident )?
+		),+
+		$(,)?
+	) => {
+		$(
+			fn $name<'scale, 'resolver>(
+				self,
+				value: $t,
+				_type_id: scale_decode::visitor::TypeIdFor<Self>,
+			) -> Result<Self::Value<'scale, 'resolver>, Self::Error> {
+				Ok(serde_json::Value::String(trimmed_hex(value$( .$convert() )?)))
+			}
+		)+
+	};
+}
+
+macro_rules! decode_to_array {
+	(
+		$(
+			$name:ident, $scale_type:ident
+		),+
+		$(,)?
+	) => {
+		$(
+			fn $name<'scale, 'resolver>(
+			self,
+			values: &mut scale_decode::visitor::types::$scale_type<'scale, 'resolver, Self::TypeResolver>,
+			_type_id: scale_decode::visitor::TypeIdFor<Self>,
+		) -> Result<Self::Value<'scale, 'resolver>, Self::Error> {
+				Ok(serde_json::Value::Array(
+					values.map(|res| {
+						res.and_then(|item| {
+							item.decode_with_visitor(self)
+						})
+					}).collect::<Result<Vec<_>,_>>()?
+				))
+			}
+		)+
+	};
+}
+
+macro_rules! decode_to_array_or_hex {
+	(
+		$(
+			$name:ident, $scale_type:ident
+		),+
+		$(,)?
+	) => {
+		$(
+			fn $name<'scale, 'resolver>(
+			self,
+			values: &mut scale_decode::visitor::types::$scale_type<'scale, 'resolver, Self::TypeResolver>,
+			_type_id: scale_decode::visitor::TypeIdFor<Self>,
+		) -> Result<Self::Value<'scale, 'resolver>, Self::Error> {
+				// Heuristic to determine if the array is a byte array:
+				// - Array must be non-empty
+				// - AND Each value decodes to an unsigned json Number (ie. not an enum)
+				// - AND each value was encoded as a single byte
+				let (is_bytes, all_values) = values.try_fold(
+					(values.remaining() > 0, Vec::with_capacity(values.remaining())),
+					|(all_bytes, mut values), res| {
+						res.and_then(|item| {
+							let decoded_item = item.decode_with_visitor(self)?;
+							let is_byte = item.bytes().len() == 1 && decoded_item.is_u64();
+							values.push(decoded_item);
+							Ok((all_bytes && is_byte, values))
+						})
+					},
+				)?;
+				if is_bytes {
+					Ok(serde_json::Value::String(untrimmed_hex(
+						all_values
+							.into_iter()
+							.map(|json| json.as_u64().unwrap() as u8)
+							.collect::<Vec<u8>>(),
+					)))
+				} else {
+					Ok(serde_json::Value::Array(all_values))
+				}
+			}
+		)+
+	};
+}
+
+impl<R: TypeResolver> scale_decode::Visitor for RawJsonVisitor<R> {
+	type Value<'scale, 'resolver> = serde_json::Value;
+	type Error = DecodeError;
+	type TypeResolver = R;
+
+	decode_to_number! {
+		visit_u8, u8,
+		visit_u16, u16,
+		visit_u32, u32,
+		visit_i8, i8,
+		visit_i16, i16,
+		visit_i32, i32,
+	}
+
+	decode_to_hex! {
+		visit_u64, u64 | to_be_bytes,
+		visit_u128, u128 | to_be_bytes,
+		visit_u256, &'scale [u8; 32],
+		visit_i64, i64 | to_be_bytes,
+		visit_i128, i128 | to_be_bytes,
+		visit_i256, &'scale [u8; 32],
+	}
+
+	decode_to_array! {
+		visit_tuple, Tuple,
+	}
+
+	decode_to_array_or_hex! {
+		visit_sequence, Sequence,
+		visit_array, Array,
```

Comment:
It's same as the difference between a vec and an array. In the former case, the length is encoded as part of the payload. In the latter case the length is known statically (ie. encoded in the metadata). 

The underlying implementation of Array and Sequence are almost identical, in fact Sequence is defined like this:

```rust
pub struct Sequence<'scale, 'resolver, R: TypeResolver> {
    bytes: &'scale [u8],
    // Mostly we just delegate to our Array logic for working with sequences.
    // The only thing we need to do otherwise is decode the compact encoded
    // length from the beginning and keep track of the bytes including that.
    values: Array<'scale, 'resolver, R>,
}
```